### ISSUE TITLE
[API] REGISTER Contact pubGruu issue

### ISSUE THREAD

**Issue by opusonline** [onsip/SIP.js#906]
Hi,

we've observed a weird behaviour with `gruu`.

**What happens:**
- The first REGISTER takes `userAgent.contact.uri` as `Contact` because no pub or temp gruu exist at that time
- The OK response includes `pub-gruu` and `temp-gruu`
- In the following REGISTER, the `Contact` uses `userAgent.contact.pubGruu`
- The OK response `Contact` is reflecting the previously used `Contact` (pubGruu) but the error is thrown `No Contact header pointing to us, dropping response`

**Description:**
- REGISTER creates Contact header with `"Contact: " + this.generateContactHeader(this.expires)` (https://github.com/onsip/SIP.js/blob/master/src/api/registerer.ts#L327)
- This uses `contact = this.userAgent.contact.toString();` (https://github.com/onsip/SIP.js/blob/b8c74916a0d6320b1137ede0c8d0afee594be9b3/src/api/registerer.ts#L652)
- Which prefers pubGruu: `contactString += this.contact.pubGruu || this.contact.uri;` (https://github.com/onsip/SIP.js/blob/b8c74916a0d6320b1137ede0c8d0afee594be9b3/src/api/user-agent.ts#L613)
- The received response is unfortunately ONLY comparing the `userAgent.contact.uri` and totally leaving the gruu contacts `if (contact.uri.user === this.userAgent.contact.uri.user) {` (https://github.com/onsip/SIP.js/blob/master/src/api/registerer.ts#L389 and https://github.com/onsip/SIP.js/blob/master/src/api/registerer.ts#L395)

**What to expect:**
Now the question comes up if this is just a bug and pubGruu should be checked as well or if there's a misunderstanding on our side?

**Notes:**
In previous versions of SIP.js the Registerer stored the very first contact and uses this again for following register requests. The current version of SIP.js is using the most recent contact for each request.

I'm looking forward for any feedback or help. Thanks in advance!

**Comment by james-criscuolo** [comment]
Can you provide logs of this occurring, as well as what version you are migrating from? Last working version would be better, but I understand if you are skipping versions.

**Comment by opusonline** [comment]
Hi,
thanks for your quick reply!

As far as I can see there was a change in your code since v0.16.0! Since then, `generateContactHeader` is using `this.userAgent.contact.toString()` but the REGISTER response updates `userAgent.contact` with `pubGruu`.
https://github.com/onsip/SIP.js/blob/0.16.0/src/api/registerer.ts#L544
https://github.com/onsip/SIP.js/blob/0.16.0/src/api/user-agent.ts#L619

All versions before are using always the same contact from the RegisterContext that was set when constructor was called.
https://github.com/onsip/SIP.js/blob/0.15.11/src/RegisterContext.ts#L207
https://github.com/onsip/SIP.js/blob/0.15.11/src/RegisterContext.ts#L460

Here's a log where you can see the error occurring. [sipjs-pubgruu-issue.log](https://github.com/onsip/SIP.js/files/6021600/sipjs-pubgruu-issue.log)

Thanks for your help!

**Comment by opusonline** [comment]
@james-criscuolo 
Any updates on this? Please let me know if you need more information. Thanks!

**Comment by opusonline** [comment]
@james-criscuolo 
A simple solution for the issue could be to add the check for `pubGruu`.
```js
if (/* existing checks */ || (this.userAgent.contact.pubGruu && equivalentURI(contact.uri, this.userAgent.contact.pubGruu);)) {
  // success
}
```
https://github.com/onsip/SIP.js/blob/master/src/api/registerer.ts#L400
https://github.com/onsip/SIP.js/blob/master/src/api/registerer.ts#L406

But I'm not sure if this is in line with the spec.

Let me know what you think. Thanks!

**Comment by james-criscuolo** [comment]
Sorry about the delay, looking to get to this in the next day or so. Wanted to ping John, as he made the change in the API transition.

**Comment by opusonline** [comment]
@james-criscuolo 
Hi. Another week has passed. Could you at least have a look or make a quick test? This is a really important part for us.
Thanks!

**Comment by james-criscuolo** [comment]
Hi @opusonline, I'm sorry but I've been unable to ping John, we've been extremely busy. Checking the RFC is the likely next step here, before we either say it should stay as is, or change it back (or change it to a third behavior, whatever matches the spec). You mentioned you were unsure if it matches the spec, figuring out the spec side would certainly save some time here.

**Comment by james-criscuolo** [comment]
Fixed with the backup solution here: https://github.com/onsip/SIP.js/commit/83417d35c9216c59c7a2f0005e6bd6359612743c

As mentioned, there is more work to be done here, as we still don't conform to spec, but this should fix this issue at least. Will close when released.

**Comment by opusonline** [comment]
@james-criscuolo Thank you so much! That makes my day 🎉

**Comment by james-criscuolo** [comment]
Released in 0.20.0.
